### ISSUE TITLE
Push doesn't work after modal dismiss

### ISSUE THREAD

**Issue by GingerBear** [wix/react-native-navigation#1670]
### Issue Description

If `showModal` is called right after a screen did mount, and after dismiss the modal, `push` on the screen doesn't work. And if I go back, all the screens that are supposed to push all get pushed.

### Steps to Reproduce / Code Snippets / Screenshots
![aug-09-2017 15-30-03](https://user-images.githubusercontent.com/902357/29140396-4f07508a-7d18-11e7-9f16-385697f94af3.gif)

to reproduce
- call `showModal` in `componentDidMount` of a screen
```js
  componentDidMount() {
    this.props.navigator.showModal({
      screen: 'example.Types.Modal'
    });
  }
```
- go to that screen 
- dismiss the modal
- try to push other screen, it won't work
- press back button, all previous screens get pushed. 

---
### Environment
* React Native Navigation version: 1.1.188
* React Native version: 0.45.1
* Platform(s) (iOS, Android, or both?): ios, not sure about android
* Device info (Simulator/Device? OS version? Debug/Release?): Simulator

**Comment by macavity23** [comment]
+1, React Native 0.45.1, RNN 1.1.198

Tabbing away to a different tab and then tabbing back also makes the 'unpushed' screens suddenly all push.

**Comment by tieusangaka** [comment]
You use dismissModal don't use dismissAllModal

**Comment by jakewtaylor** [comment]
Any solutions to this?

Getting this problem with `dismissModal()` **not** `dismissAllModals()`.

My script:
```JavaScript
const { navigator } = this.props;
navigator.dismissModal(); // works fine
navigator.push({ ... }; // doesn't work
```

**Comment by yusufyildirim** [comment]
I'm doing the same thing as @GingerBear 's answer but this is not a healthy solution. This is big problem guys, can anyone explain and solve this?

**Comment by leolusoli** [comment]
+1

**Comment by drpiou** [comment]
I don't know if it is completely related, but when I do a `dismissAllModals()`,  `pop()` stop not working on iOS (Android is fine).

**Comment by spencewine** [comment]
I ran into a similar situation with using this.props.navigator.push. I was able to solve it by using async await.

**Comment by kyle-ssg** [comment]
+1 this also seems to happen when used alongside libs with native modals (e.g. facebook accountkit)

**Comment by jasonmerino** [comment]
I too am seeing this when calling `navigator.handleDeepLink` and trying to handle the deep link. It seems to only have shown up when I added navigation event listeners to root screens in multiple tabs. I got around it with a combination of `setTimeout`, `setImmediate` (for switching from one tab to the actual one I'm trying to switch to) and deferring the `navigator.dismissModal` call until after all the underlying routing is completed. It's definitely not a pretty piece of code but it works for the time being until this gets sorted out/work on v2 is completed. 

Thanks for all your work on this Wix team! On the whole this is a pretty killer library.

**Comment by Brunoalcau** [comment]
+1

**Comment by getaaron** [comment]
I can also reproduce this when calling `navigator.push` in JavaScript after dismissing a modal using in Swift code.

The issue is resolved if I wait for the `dismiss(animated:completion:)` completion handler to get called before calling `navigator.push`.

This probably isn’t a great workaround for most use cases, but I mention it in case it helps debug the issue.

**Comment by kkgelu** [comment]
@kyle-ssg can confirm that if any modal(`navigator.showModa`l) is presented, the native modal such as `AccountKit.loginWithPhone` won't work/show. Any workaround?

**Comment by kyle-ssg** [comment]
@kkgelu In my case my issue was along the lines of what @getaaron was facing, his solution was similar to mine

**Comment by dev6james** [comment]
I've had the same behaviour.

Is this planned to be fixed in v1? Does anyone who's used v2 know if it's also present?

It sounds like from @GingerBear's description (That it happens if showModal is triggered in a didMount handler), that @getaaron's fix wouldn't solve that case because the implication is that 'dismiss(animated:completion:)' never fires for some reason?

**Comment by ppozniak** [comment]
Same issue here, but only on iOS. I thought I would get away by using deep linking, but this doesn't work :(

So I got away from this by doing such thing
```
// inside of button handler
this.props.navigator.handleDeepLinking({ ... }); // first call handleDeep link
this.props.navigator.dismissModal(); // then AFTER  dismissModal
```
And on other screen normal deep linking handling where i just `push`.

Dismissing modal inside of deep linking handler didn't work.

**Comment by vysotsky** [comment]
The same issue. And only on iOS.

**Comment by mobitar** [comment]
I found out what the issue was:

In `RCCNavigationController`, the method `- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated` (line 451 in my fork), looks like this:

```
- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated
{
  if(_transitioning)
  {
    NSDictionary *pushDetails =@{ @"viewController": viewController, @"animated": @(animated) };
    [_queuedViewControllers addObject:pushDetails];

    return;
  }

  _transitioning = YES;
  
  [super pushViewController:viewController animated:animated];
}
```

What's happening with this bug is that `_transitioning` is never reset to false, for whatever reason. So, when you go to push, it's queued, but never run.

The solution I've found that works is:

```
- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated
{
  if(_transitioning)
  {
    NSDictionary *pushDetails =@{ @"viewController": viewController, @"animated": @(animated) };
    [_queuedViewControllers addObject:pushDetails];

    return;
  }

  _transitioning = YES;
  
  [CATransaction begin];
  [CATransaction setCompletionBlock:^(){
    _transitioning = NO;
  }];

  [super pushViewController:viewController animated:animated];
  
  [CATransaction commit];
}
```

Basically, since push uses CoreAnimation, we can use CATransaction's completion handler to reset our transitioning flag.

(This is a functional fix. However, the real underlying issue is why `- (void)navigationController:(UINavigationController *)navigationController didShowViewController:(UIViewController *)viewController animated:(BOOL)animated` is not called in some cases, which is where `_transitioning` is supposed to be reset in the first place.)

**Comment by jmknoll** [comment]
+1 to bring more attention to this issue, and add a little bit more info.

If I delay the dismissModal() call until after the push is complete, it dismisses the modal as well as the screen that I just pushed. 

None of the solutions above have worked for me.

For the creators of this library, it is an excellent library, but this is a pretty significant problem that (hopefully) should be addressed. Is there a plan to fix this in v1? Will migrating to v2 address this issue?

**Comment by sapjax** [comment]
In v1 modalPresentationStyle: 'overFullScreen'  will cause this issue.

**Comment by doublex** [comment]
+1

**Comment by dieunt** [comment]
The same issue.

**Comment by andrebraga** [comment]
Stale only because nobody managed to solve it! Reopen please.

**Comment by andrebraga** [comment]
By the way this is literally _the issue_ that makes this library unusable to me. That, and the fact that you haven't implemented popToRoot.

**Comment by diegolmello** [comment]
Hey, guys.
In my case, I open a modal and, depending on what the user'll do, I may have to close it and push a screen.
For example, the user opens a modal in order to create a conversation with others.
He picks some of them and hit save.
My app closes the modal and push the stack to the new channel.
This what should be happening.

What is happening is after closing the modal, the app is stuck on the previous screen (without pushing to the new channel).
After debugging, I notice that `Screen.js` uses it's context to action (push, pop, dismissModal, etc).
The problem is that it's context have changed, since push action needs to be executed from previous stack view instead of modal stack.
So, dismiss action runs fine, but push doesn't work, because the modal doesn't exists anymore.

That said, I need to say I'm aware of v2 and the new API.
For those who haven't tried/read the docs, do so.
The new API allows you to set what component is performing the action.
In my case, v2 would works fine, because it's just a matter of setting previous stack as performer of `push new channel` action.
There's a few reasons why I am using v1, but I won't enter in details in the issue (I talked about in another issue).

Since I can't set origin component, I can "hack" the method responsible for finding what component should execute that action.
This is `RCCManager.getControllerWithId` does.
It's a very simple method that find the component by `id` and the `componentType`.

```objective-c
-(id)getControllerWithId:(NSString*)componentId componentType:(NSString*)componentType {
    if (componentId == nil) {
        return nil;
    }
    
    id component = nil;
    
    NSMutableDictionary *componentsDic = self.modulesRegistry[componentType];
    if (componentsDic != nil) {
        component = componentsDic[componentId];
    }
    
    return component;
}
```
This is my version:
```objective-c
-(id)getControllerWithId:(NSString*)componentId componentType:(NSString*)componentType {
    if (componentId == nil) {
        return nil;
    }
    
    id component = nil;
    
    NSMutableDictionary *componentsDic = self.modulesRegistry[componentType];
    if (componentsDic != nil) {
        if ([componentType isEqualToString:@"NavigationControllerIOS"]) {
            component = [[componentsDic allValues] lastObject];
        } else {
            component = componentsDic[componentId];
        }
    }
    
    return component;
}
```

What I'm doing is changing the search to always look for the last `NavigationControllerIOS` on the dictionary.
When the modal is dismissed, this function will search for the stack active, find the previous stack and push the new view correctly.

Use this at you're own risk, but it's working fine in my case.
If I find something breaking, I'll edit this comment.
Let me know if this works for you guys too.

**Comment by andrebraga** [comment]
The order in a dictionary is never guaranteed. The fact that your code works is rather random to be honest and quite prone to change between OS versions as the hashing algorithm may change (as it just did in Swift 4.2).

What you're looking for instead is filtering that dictionary for navigators, sorting the resulting array by id (assuming they sort in monotonically increasing order) and then grabbing the last one, or quite possibly the last viewController in that navigator's stack.

I'm not working with RN code at the moment so it will take me a while to test this. If you manage to get there before me by all means let us know.

**Comment by diegolmello** [comment]
@andrebraga Yes, that's why didn't opened a PR.
As I said, it's a hack, but it can help other people.

**Comment by ammichael** [comment]
For the record, I'm also having this exactly same issue. Whenever I'm pushing after dismissing a modal or lightbox, it just stop working. Right now I'm setting timeouts, but sometimes it fails as well (maybe setting a higher timeout could do it).

It's been more than 1 year since the first report of this issue, and I seriously doubt it will be fixed in v1. I'll try @diegolmello solution, but right now the best solution is setting timeouts when showing modals and pushing routes.

**Comment by geastwood** [comment]
We have the same issue on Android, by calling `destroy()` directly instead of `animationHide()` in LightBox.java::hide method fixed the issue, however of course you lose the animation on lightbox dismiss.

**Comment by jakewtaylor** [comment]
@guyca can you clarify why you've closed this? AFAIK this hasn't been fixed, seems strange to close without any explanation.

**Comment by guyca** [comment]
Hey @jakewtaylor 
Sorry for that, I closed all issues with the v1 label. v1 isn't supported and therefore there's no point in keeping these issues open.

**Comment by amrlabib** [comment]
Using `animationType: 'none'` helped me incase someone is still having this issue in v1 
```
Navigation.dismissModal({ animationType: 'none' });
```
